### PR TITLE
[SPARK-36098][CORE] Grouping exception in core/storage

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?
This PR group exception messages in core/src/main/scala/org/apache/spark/storage

### Why are the changes needed?
It will largely help with standardization of error messages and its maintenance.

### Does this PR introduce _any_ user-facing change?
No. Error messages remain unchanged.

### How was this patch tested?
No new tests - pass all original tests to make sure it doesn't break any existing behavior.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+package org.apache.spark.errors
+
+import java.io.IOException
+
+import org.apache.spark.SparkException
+import org.apache.spark.shuffle.{FetchFailedException, ShuffleManager}
+import org.apache.spark.storage.{BlockId, BlockManagerId, BlockNotFoundException, UnrecognizedBlockId}
+
+object SparkCoreErrors {
+  def failedParseBlockIdError(name: String): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:27]: unrecognizedBlockIdError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+  def taskHasNotLockedBlockError(currentTaskAttemptId: Long, blockId: BlockId): Throwable = {
+    new SparkException(s"Task $currentTaskAttemptId has not locked block $blockId for writing")
+  }
+
+  def blockDoesNotExistError(blockId: BlockId): Throwable = {
+    new SparkException(s"Block $blockId does not exist")
+  }
+
+  def cannotSaveBlockOnDecommissionExecutorError(blockId: BlockId): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:39]: cannotSaveBlockOnDecommissionedExecutorError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    new SparkException(s"Block $blockId cannot be saved on decommissioned executor")
+  }
+
+  def waitingForReplicationFinishError(t : scala.Throwable): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:43]: `waitingForReplicationToFinishError(e: Throwable)`
nit: remove the white space after `t`

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    new SparkException("Error occurred while waiting for replication to finish", t)
+  }
+
+  def unableRegisterWithExternalShuffleServerError(t : scala.Throwable): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:47]: unableToRegisterWithExternalShuffleServerError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    new SparkException(s"Unable to register with external shuffle server due to : ${t.getMessage}",
+      t)
+  }
+
+  def waitingForAsyncReregistrationError(t : scala.Throwable): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:52]: nit: extra white space

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  def unexpectedShuffleBlockWithUnsupportedResolverError(
+      shuffleManager: ShuffleManager,
+      blockId: BlockId): Throwable = {
+    new SparkException(s"Unexpected shuffle block ${blockId} with unsupported shuffle " +
+      s"resolver ${shuffleManager.shuffleBlockResolver}")
+  }
+
+  def failedStoreBlockOnBlockManagerError(
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:63]: failToStoreBlockOnBlockManagerError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      blockManagerId: BlockManagerId,
+      blockId: BlockId): Throwable = {
+    new SparkException(s"Failure while trying to store block $blockId on $blockManagerId.")
+  }
+
+  def readLockedBlockNotFoundError(blockId: BlockId): Throwable = {
+    new SparkException(s"Block $blockId was not found even though it's read-locked")
+  }
+
+  def failedBlockGetEvenHeldALockError(blockId: BlockId): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:73]: failToGetBlockWithLockError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+  def interruptedError(): Throwable = {
+    new InterruptedException()
+  }
+
+  def blockStatusQueryReturnedNullError(blockId: BlockId): Throwable = {
+    new SparkException(s"BlockManager returned null for BlockStatus query: $blockId")
+  }
+
+  def unexpectedResultBlockManagerMasterEndpointError(): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:89]: unexpectedBlockManagerMasterEndpointResultError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    new SparkException("BlockManagerMasterEndpoint returned false, expected true.")
+  }
+
+  def failedCreateDirectoryError(path: String, maxAttempts: Int): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:93]: failToCreateDirectoryError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+      shuffleId: Int,
+      mapId: Long,
+      mapIndex: Int,
+      reduceId: Int,
+      message: String,
+      cause: Throwable = null): Throwable = {
+    new FetchFailedException(bmAddress, shuffleId, mapId, mapIndex, reduceId, message, cause)
+  }
+
+  def failedGetNonShuffleBlockError(blockId: BlockId, e: Throwable): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:117]: failToGetNonShuffleBlockError

**Review by allisonwang-db — COMMENTED** [review]
Looks good. Please also rebase the latest master.

**Comment by dgd-contributor** [comment]
@allisonwang-db I've rebase and resolve all of  your suggestions. Can you re-review.

**Review by allisonwang-db — APPROVED** [review]

**Comment by dgd-contributor** [comment]
cc @cloud-fan

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 11]
```diff
@@ -270,7 +271,7 @@ private[spark] class BlockManager(
     // Don't reject broadcast blocks since they may be stored during task exec and
     // don't need to be migrated.
     if (isDecommissioning() && !blockId.isBroadcast) {
-        throw new BlockSavedOnDecommissionedBlockManagerException(blockId)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:273]: I think the principle is we should not change the exception class?

Comment (dgd-contributor) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:273]: this exception class is only used once, I don't know should we keep it or not?

Comment (allisonwang-db) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:273]: Good catch. Yes let's keep the original exception classes and avoid changing them during the grouping process. We can make separate PRs in the future to improve quality.

Comment (dgd-contributor) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:273]: Done!

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**